utp-testing/src/main.rs

```diff
@@ -0,0 +1,147 @@
+use discv5::Discv5Event;
+use log::debug;
+use std::str::FromStr;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+use trin_core::portalnet::discovery::Discovery;
+use trin_core::portalnet::types::messages::{PortalnetConfig, ProtocolId};
+use trin_core::portalnet::Enr;
+use trin_core::utp::stream::UtpListener;
+use trin_core::utp::trin_helpers::{UtpMessage, UtpMessageId};
+
+pub struct TestApp {
+    utp_listener: Arc<RwLock<UtpListener>>,
+}
+
+impl TestApp {
+    async fn send_utp_request(&mut self, connection_id: u16, payload: Vec<u8>, enr: Enr) {
+        self.utp_listener
+            .write()
+            .await
+            .listening
+            .insert(connection_id, UtpMessageId::OfferAcceptStream);
+
+        let mut conn = self
+            .utp_listener
+            .write()
+            .await
+            .connect(connection_id, enr.node_id())
+            .await
+            .unwrap();
+
+        let mut buf = [0; 1500];
+        conn.recv(&mut buf).await.unwrap();
+
+        conn.send_to(&UtpMessage::new(payload.clone()).encode()[..])
+            .await
+            .unwrap();
+
+        tokio::spawn(async move {
+            conn.close().await.unwrap();
+            debug!("connecton state: {:?}", conn.state)
+        });
+    }
+
+    async fn process_utp_request(&self) {
+        let mut event_stream = self
+            .utp_listener
+            .read()
+            .await
+            .discovery
+            .discv5
+            .event_stream()
+            .await
+            .unwrap();
+
+        let listener = Arc::clone(&self.utp_listener);
+
+        tokio::spawn(async move {
+            while let Some(event) = event_stream.recv().await {
+                let request = match event {
+                    Discv5Event::TalkRequest(r) => r,
+                    _ => continue,
+                };
+
+                let protocol_id =
+                    ProtocolId::from_str(&hex::encode_upper(request.protocol())).unwrap();
+
+                if let ProtocolId::Utp = protocol_id {
+                    listener.write().await.process_utp_request(request).await;
+                    listener.write().await.process_utp_byte_stream().await;
+                };
+            }
+        });
+    }
+
+    async fn prepare_to_receive(&self, connection_id: u16) {
+        // listen for incoming connection request on conn_id, as part of utp handshake
+        self.utp_listener
+            .write()
+            .await
+            .listening
+            .insert(connection_id, UtpMessageId::OfferAcceptStream);
+
+        // also listen on conn_id + 1 because this is the actual receive path for acceptor
+        self.utp_listener
+            .write()
+            .await
+            .listening
+            .insert(connection_id + 1, UtpMessageId::OfferAcceptStream);
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    tracing_subscriber::fmt::init();
+
+    let client_port = 9002;
```

Comment:
Might not be a big issue for now - but this will probably fail if those ports are not available.

Reply:
Yeah, this is just a init way to test manually simple uTP to uTP connection. This framework will evolve a lot and will use docker containers in the future.